### PR TITLE
Train End Error Handling Fix

### LINKED ISSUE #5766: Errors within try/except of train(self) are misrepresented as checkpointing MisconfigurationException
## 🐛 Bug

I think I found a bug, where errors probably caused by users are misrepresented as checkpointing MisconfigurationException even though the checkpointing is configured correctly.

This happens when errors are raised within training (such as `RuntimeErrors` or CUDA-OOM errors) and bubble up to the `try/except` command in `def train(self)` function, which can be found here: https://github.com/PyTorchLightning/pytorch-lightning/blob/65247ec47cbe8f0254f444f409025585806f9113/pytorch_lightning/trainer/trainer.py#L550
As these errors are not excepted here, the code continues to execute the code in `finally`. This calls `self.train_loop.on_train_end()` which proceeds to save a checkpoint. If one monitors on a _validation_ metric, such as `val/accuracy`, no value has been saved as the error occured during the _training_. Thus, in turn a `MisconfigurationException` is raised by the checkpointing code, stating that the monitored metric is not found in the returned metrics.

### To Reproduce

- configure checkpointing to monitor a validation metric, which (by definition) must not be logged during training.
- raise any error during training, such as `RuntimeError` but not `KeybordInterrupt` (that is the only one which is caught).

### Expected behavior

The error should not be _glossed over_ by the `finally` statement and be raised all the way to the top, so the user can see it and fix the underlying bug.

### Environment

- OS: Linux
- env-type: conda
- pytorch-lightning 1.1.4
- pytorch 1.7.0
- python 3.8.6
 - CUDA/cuDNN version: 11.0
 - GPU models and configuration: One v100, 32GB vRAM

### Lazy patch:
A very hacky way of at least letting the user know about their error, is to modify train such that after `except KeyboardInterrupt` all other errors are caught with `except Exception as e` and then immediately printing the error with `print(e)`. Unfortunately raising the error with `raise e` does not work because the `finally` code is executed first, raising its own `MisconfigurationException`.

The full code of `train` would look like this (changes are below the excepted `KeyboardInterrupt`):
```python
    def train(self):
        self.run_sanity_check(self.get_model())

        # set stage for logging
        self.logger_connector.set_stage("train")

        self.checkpoint_connector.has_trained = False

        # enable train mode
        model = self.get_model()
        model.train()
        torch.set_grad_enabled(True)

        # reload data when needed
        self.train_loop.reset_train_val_dataloaders(model)

        # hook
        self.train_loop.on_train_start()

        try:
            if self.train_loop.should_skip_training():
                return
            # run all epochs
            for epoch in range(self.current_epoch, self.max_epochs):

                # hook
                self.train_loop.on_train_epoch_start(epoch)

                with self.profiler.profile("run_training_epoch"):
                    # run train epoch
                    self.train_loop.run_training_epoch()

                if self.max_steps and self.max_steps <= self.global_step:
                    return

                # update LR schedulers
                self.optimizer_connector.update_learning_rates(interval='epoch')

                # early stopping
                met_min_epochs = epoch >= self.min_epochs - 1
                met_min_steps = self.global_step >= self.min_steps if self.min_steps else True

                if self.should_stop:
                    if met_min_epochs and met_min_steps:
                        return
                    log.info(
                        'Trainer was signaled to stop but required minimum epochs'
                        f' ({self.min_epochs}) or minimum steps ({self.min_steps}) has'
                        ' not been met. Training will continue...'
                    )

        except KeyboardInterrupt:
            rank_zero_warn('Detected KeyboardInterrupt, attempting graceful shutdown...')

            # user could press ctrl+c many times... only shutdown once
            if not self.interrupted:
                self.interrupted = True
                self._state = TrainerState.INTERRUPTED
                self.on_keyboard_interrupt()
        except Exception as e:
            print(e)  # better would be raising it, but that would be executed after the finally
        finally:
            # hook
            self.train_loop.on_train_end()
```

### Test
A test to catch this problem in the future would be to checkpoint on a validation metric, raise any error in the training (except `KeyboardInterrupt` which is excepted) and assert that it bubbles up all the way.

### LINKED ISSUE #6807: Runtime- and Assertionerror handling in trainer.run_train
## 🐛 Bug
Runtime- (e.g. Out of Memory Errors) /Assertionerror are ignored when running `trainer.fit(...)`. The error will not be raised correctly and the script will continue.
This can waste a lot of time in some cases:

```python
# prepare code
...

# train (model raises an error, e.g. Out of Memory which is not raised by trainer)
trainer.fit(...)

# will continue here
# time intensive computation / evaluation / prediction
...
```
 While the error is printed correctly, it is not raised and thus the script will continue.
 https://github.com/PyTorchLightning/pytorch-lightning/blob/bb9ace43334ad50e3758d9cff08ad34216c7d4da/pytorch_lightning/trainer/trainer.py#L621-L634
 

### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

### Expected behavior
The script should stop after the trainer cleaned up when an Assertion or Runtime error occurs during training.

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### Additional context
Simply saving the exception and raising it after the trainer called the final hook should be sufficient.

### LINKED ISSUE #6842: Trainer Error Handling Fix
## 🐛 Bug

For distributed training, if a subset of ranks fail during some training step, the current setting tries to gracefully shutdown by calling 

https://github.com/PyTorchLightning/pytorch-lightning/blob/22a266d8b8cf57455cc863e20491e416ec635ba7/pytorch_lightning/trainer/trainer.py#L634

However, as not all ranks enter this `on_train_end`, we have the logic to perform model checkpoint which would hang while broadcasting.

https://github.com/PyTorchLightning/pytorch-lightning/blob/22a266d8b8cf57455cc863e20491e416ec635ba7/pytorch_lightning/callbacks/model_checkpoint.py#L723-L730

see also discussion #6807 and #6791

### Pitch

keep special case for `KeyboardInterrupt` expection, for the rest exceptions, we raise the exception and remove finally

```
  try:
      ..... (run training epochs)
      # hook
      self.train_loop.on_train_end()
  except KeyboardInterrupt:
      rank_zero_warn('Detected KeyboardInterrupt, attempting graceful shutdown...')
      # user could press Ctrl+c many times... only shutdown once
      if not self.interrupted:
          self.state = TrainerState.INTERRUPTED
          self.on_keyboard_interrupt()
          self.train_loop.on_train_end()
  except:
      print_exc()
      raise
```

### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by shuyingsunshine21**
## What does this PR do?

This PR changes the logic so we NO LONGER try to save on `KeyboardInterrupt` or call callback's `on_train_end`

Fixes #6842
Fixes #6807
Fixes #5766

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/optimization/test_manual_optimization.py [hunk 1]
```diff
@@ -724,10 +727,6 @@ def optimizer_closure():
             weight_after = self.layer.weight.clone()
             assert not torch.equal(weight_before, weight_after)
 
-        def training_epoch_end(self, outputs) -> None:
-            # outputs should be an array with an entry per optimizer
-            assert len(outputs) == 2
-
```
Comment (shuyingsunshine21) [inline tests/trainer/optimization/test_manual_optimization.py:730]: removing these, as we have `model.training_epoch_end = None` override in L737

File: tests/trainer/optimization/test_manual_optimization.py [hunk 2]
```diff
@@ -68,8 +68,9 @@ def training_step(self, batch, batch_idx):
             assert torch.all(self.layer.weight.grad == 0)
 
         def training_epoch_end(self, outputs) -> None:
-            # outputs should be an array with an entry per optimizer
-            assert len(outputs) == 2
+            # outputs is empty as training_step does not return
+            # and it is not automatic optimization
+            assert len(outputs) == 0
```
Comment (shuyingsunshine21) [inline tests/trainer/optimization/test_manual_optimization.py:74]: removing the handling reveals this test errors, like manual optimization (when training step does not return).

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -615,6 +616,7 @@ def run_train(self) -> None:
 
                 if self.should_stop:
                     if met_min_epochs and met_min_steps:
+                        self.train_loop.on_train_end()
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:619]: shall we also remove this from `train_loop.on_train_end` ? https://github.com/PyTorchLightning/pytorch-lightning/blob/b7a22ba046ba57072a71b12d16caff000e66f798/pytorch_lightning/trainer/training_loop.py#L114-L118

as this is one of the possible sources of bugs with the attempt at graceful shutdown

File: tests/trainer/optimization/test_manual_optimization.py (hunk 2 above)
Comment (ananthsub) [reply tests/trainer/optimization/test_manual_optimization.py:74]: great find! this broad try/catch was suppressing existing errors in our tests

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -420,7 +420,7 @@ def fit(
         # we reuse fit for other functions. When already set, it shouldn't be modified.
         if not self.state.running:
             self.state = TrainerState.FITTING
-        if self._running_stage is None:
+        if self._running_stage is None or self._running_stage == RunningStage.TUNING:
```
Comment (shuyingsunshine21) [inline pytorch_lightning/trainer/trainer.py:423]: this is for test failure `trainer.tune(model)` here
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/trainer/test_states.py#L45-L48

**Review by tchaton — APPROVED** [review]
LGTM ! Nice catches !

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:619]: Yes. I have a PR in progress for this. Need to finish it. This is part of what causes https://github.com/PyTorchLightning/pytorch-lightning/issues/6470

**Comment by awaelchli** [comment]
looks great, is it still wip or already for review? :)

File: pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:423]:
```suggestion
        if self._running_stage is None or self.tuning:
```

The tuning case is different because it is treated as training internally, we might want to set the stage back to tuning at the end of this, if tuning was initially set

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -633,14 +635,15 @@ def run_train(self) -> None:
             if not self.interrupted:
                 self.state = TrainerState.INTERRUPTED
                 self.on_keyboard_interrupt()
-        except (RuntimeError, AssertionError):
-            # if an exception is raised, the finally block is executed and can hide the actual exception
-            # that was initially raised if `on_train_end` also raises an exception. we want to avoid that
-            # for assertions and other runtime errors so we aren't misled while debugging
+                self.accelerator.on_train_end()
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:638]: Why do we use `accelerator.on_train_end` instead of `train_loop` here and below?

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -633,14 +635,15 @@ def run_train(self) -> None:
...
+                self._running_stage = None
+        except:
+            # give accelerators a chance to finish
+            self.accelerator.on_train_end()
+            # reset bookkeeping
+            self._running_stage = None
             print_exc()
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:645]: Probably don't need this print now that we raise again

Comment (shuyingsunshine21) [reply pytorch_lightning/trainer/trainer.py:645]: Thanks @carmocca, @awaelchli  and @tchaton for the review, will address the comments and publish.

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (shuyingsunshine21) [reply pytorch_lightning/trainer/trainer.py:619]: @ananthsub , for this case, train does not have error (just early stopping), should we run `train_loop.on_train_end` to perform checkpointing?

@camruta , for why here I use `train_loop.on_train_end`, but for the case exception, I use `accelerator.on_train_end`, is because trying to avoid checkpointing when there is error for some ranks but not all (hence hang in broadcast).


To avoid divergence (normal end vs. error end), do you think we actually should remove the problematic checkpointing logic in `training_loop.on_train_end`. But there is also risk of adding logics in `training_loop.on_train_end` which could potentially need all ranks to be involved. How do we enforce that.

Comment (shuyingsunshine21) [reply pytorch_lightning/trainer/trainer.py:618]: open to discussion

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:637]: train_loop.on_train_end would call accelerator.on_train_end internall also, but I suspect it is removed because they want to avoid checkpointing on this interruption?

Comment (shuyingsunshine21) [reply pytorch_lightning/trainer/trainer.py:637]: Yes, the reason is avoid the "troublesome" checkpointing (when there is error for some ranks but not all, hence hang in broadcast).

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:637]: or the case where we fail during training, but the checkpoint callback is configured for a monitor available only in validation

**Comment by shuyingsunshine21** [comment]
> I assume this PR will also close #6807, correct?

I think so.

**Comment by carmocca** [comment]
Labeling as API change since this removes support for saving checkpoints on Ctrl+C/exception and it might not be okay for some users. I'd like to have @williamFalcon's approval first.

**Comment by carmocca** [comment]
Also given the current changes, this should be updated since callback's `on_train_end` wont be called on Ctrl+C

https://github.com/PyTorchLightning/pytorch-lightning/blame/master/docs/source/common/trainer.rst#L145-L149

**Review by williamFalcon — APPROVED** [review]

**Review by Borda — COMMENTED** [review]
can we have a Traner argument that would enable all these emergency savings, I this was quite a useful feature in case you run training for a while and then timeout or any other kill came...

**Comment by shuyingsunshine21** [comment]
> can we have a Traner argument that would enable all these emergency savings, I this was quite a useful feature in case you run training for a while and then timeout or any other kill came...

wonder if setting "frequent" checkpointing in training would help.

**Comment by shuyingsunshine21** [comment]
looks like the test fail is related to a recent test: https://github.com/PyTorchLightning/pytorch-lightning/pull/6969/files

because of the current error handling logic, it does not catch this assertion error


https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/trainer/test_training_loop.py#L140-L143

File: tests/trainer/test_training_loop.py [hunk 7]
```diff
@@ -132,15 +141,16 @@ def test_outputs_format(tmpdir):
 
     class HookedModel(BoringModel):
         def training_step(self, batch, batch_idx):
-            self.log("foo", "bar")
-            return super().training_step(batch, batch_idx)
+            output = super().training_step(batch, batch_idx)
+            self.log("foo", 123)
+            output["foo"] = 123
+            return output
```
Comment (shuyingsunshine21) [inline tests/trainer/test_training_loop.py:147]: test fix for not captured before.

Comment (ethanwharris) [reply tests/trainer/test_training_loop.py:147]: Yeah that's my bad, fix looks good 👍

Comment (shuyingsunshine21) [reply tests/trainer/test_training_loop.py:147]: no problem :)

**Review by justusschock — APPROVED** [review]

**Comment by tchaton** [comment]
Dear @shuyingsunshine21, 

You have 2 failing test on Azure. Do you need help to solve this ?

Best,
T.C

**Comment by shuyingsunshine21** [comment]
> Dear @shuyingsunshine21,
> 
> You have 2 failing test on Azure. Do you need help to solve this ?
> 
> Best,
> T.C

Oh, thanks, did not notice that. Let me take a look.

**Comment by carmocca** [comment]
> Oh, thanks, did not notice that. Let me take a look.

@shuyingsunshine21 The issue is on our end. Seems like there was a fairscale update that is breaking our CI. We are taking a look.
